File: generate_questions.js
// generate_questions.js
const { GoogleGenerativeAI } = require("@google/generative-ai");
const fs = require('fs');
const path = require('path');

// --- CONFIGURATION ---
// ideally load from environment variable: process.env.GEMINI_API_KEY
const API_KEY = "AIzaSyDz9GHDo2CppUkkx_06h4eUQeq9nXbQrBY"; // Replace with your new key after revoking the old one
const genAI = new GoogleGenerativeAI(API_KEY);

// The model to use (Flash is fast and cheap, Pro is smarter)
const model = genAI.getGenerativeModel({ model: "gemini-2-flash" });

const REGIONS = [
    { id: "ttah", ar: "طنجة تطوان الحسيمة", en: "Tanger-Tetouan-Al Hoceima" },
    { id: "loriental", ar: "الشرق", en: "Oriental" },
    { id: "fm", ar: "فاس مكناس", en: "Fes-Meknes" },
    { id: "rsk", ar: "الرباط سلا القنيطرة", en: "Rabat-Sale-Kenitra" },
    { id: "bmk", ar: "بني ملال خنيفرة", en: "Beni Mellal-Khenifra" },
    { id: "cs", ar: "الدار البيضاء سطات", en: "Casablanca-Settat" },
    { id: "ms", ar: "مراكش آسفي", en: "Marrakesh-Safi" },
    { id: "dt", ar: "درعة تافيلالت", en: "Draa-Tafilalet" },
    { id: "sm", ar: "سوس ماسة", en: "Souss-Massa" },
    { id: "gon", ar: "كلميم واد نون", en: "Guelmim-Oued Noun" },
    { id: "lseh", ar: "العيون الساقية الحمراء", en: "Laayoune-Sakia El Hamra" },
    { id: "dod", ar: "الداخلة وادي الذهب", en: "Dakhla-Oued Ed-Dahab" }
];

const OUTPUT_FILE = path.join(__dirname, 'data', 'questions.json');

async function generateQuestionsForRegion(region) {
    console.log(`🚀 Generating questions for: ${region.ar}...`);

    const prompt = `
You are an educational AI generating **Moroccan regional quizzes** in Arabic.
Follow these instructions **strictly**:

1. Generate exactly **30 questions** for the region: **${region.ar} (${region.en})**.
2. Questions must cover these domains:
   - Geography (location, borders, coordinates)
   - History (important events, independence, key figures)
   - Culture (festivals, food, music, dialects)
   - Landmarks / monuments / tourism
   - Major cities and economy
   - Fun facts & trivia
3. Each question object must have:
   - "id": unique string (e.g., "${region.id}_Q1")
   - "question": Arabic text
   - "options": array of 4 strings in Arabic
   - "answer": exact string matching the correct option
4. Return **ONLY a valid JSON array**. No markdown, no explanation.

Example format:
[
  {
    "id": "${region.id}_Q1",
    "question": "ما هي عاصمة منطقة ${region.ar}؟",
    "options": ["طنجة", "فاس", "الرباط", "مراكش"],
    "answer": "طنجة"
  }
]

Now, generate the 30 questions for ${region.ar}.
`;

    try {
        const result = await model.generateContent(prompt);
        const response = await result.response;
        let text = response.text();

        // Clean up potential markdown code blocks if the model adds them
        text = text.replace(/```json/g, '').replace(/```/g, '').trim();

        // Parse to ensure valid JSON
        const questions = JSON.parse(text);

        // Tag with region ID (double safety)
        const taggedQuestions = questions.map((q, i) => ({
            ...q,
            id: q.id || `${region.id}_Q${i+1}`, // Ensure ID exists
            region_id: region.id
        }));

        console.log(`✅ Success: Generated ${taggedQuestions.length} questions for ${region.ar}`);
        return taggedQuestions;

    } catch (error) {
        console.error(`❌ Error generating for ${region.ar}:`, error.message);
        return []; // Return empty on failure so the script continues
    }
}

async function main() {
    let allQuestions = [];

    for (const region of REGIONS) {
        const qs = await generateQuestionsForRegion(region);
        allQuestions = allQuestions.concat(qs);

        // Small delay to avoid rate limiting
        await new Promise(resolve => setTimeout(resolve, 1000));
    }

    // Ensure data directory exists
    const dataDir = path.dirname(OUTPUT_FILE);
    if (!fs.existsSync(dataDir)) {
        fs.mkdirSync(dataDir, { recursive: true });
    }

    // Write to file
    fs.writeFileSync(OUTPUT_FILE, JSON.stringify(allQuestions, null, 2), 'utf-8');
    console.log(`\n🎉 ALL DONE! Total questions saved: ${allQuestions.length}`);
    console.log(`📁 File location: ${OUTPUT_FILE}`);
}

main();

File: generate_questions.py
import json
import os
import time
import urllib.error
import urllib.request

# --- CONFIGURATION ---
# Paste your Gemini API Key here
API_KEY = "YOUR_GEMINI_API_KEY"
MODEL = "gemini-1.5-flash"
URL = f"https://generativelanguage.googleapis.com/v1beta/models/{MODEL}:generateContent?key={API_KEY}"

REGIONS = [
    {"id": "MA-01", "ar": "طنجة تطوان الحسيمة", "en": "Tanger-Tetouan-Al Hoceima"},
    {"id": "MA-02", "ar": "الشرق", "en": "Oriental"},
    {"id": "MA-03", "ar": "فاس مكناس", "en": "Fes-Meknes"},
    {"id": "MA-04", "ar": "الرباط سلا القنيطرة", "en": "Rabat-Sale-Kenitra"},
    {"id": "MA-05", "ar": "بني ملال خنيفرة", "en": "Beni Mellal-Khenifra"},
    {"id": "MA-06", "ar": "الدار البيضاء سطات", "en": "Casablanca-Settat"},
    {"id": "MA-07", "ar": "مراكش آسفي", "en": "Marrakesh-Safi"},
    {"id": "MA-08", "ar": "درعة تافيلالت", "en": "Draa-Tafilalet"},
    {"id": "MA-09", "ar": "سوس ماسة", "en": "Souss-Massa"},
    {"id": "MA-10", "ar": "كلميم واد نون", "en": "Guelmim-Oued Noun"},
    {"id": "MA-11", "ar": "العيون الساقية الحمراء", "en": "Laayoune-Sakia El Hamra"},
    {"id": "MA-12", "ar": "الداخلة وادي الذهب", "en": "Dakhla-Oued Ed-Dahab"},
]


def generate_for_region(region):
    print(f"Generating for {region['ar']}...")

    prompt = f"""
    You are an educational AI. Generate 3 questions in Arabic about the Moroccan region: {region["ar"]} ({region["en"]}).
    Topics: Geography, History, Culture.
    Return ONLY a valid JSON array.
    Schema: [{{ "id": "{region["id"]}_Q1", "question": "...?", "options": ["...", "...", "...", "..."], "answer": "...", "region_id": "{region["id"]}" }}]
    """

    payload = {"contents": [{"parts": [{"text": prompt}]}]}

    data = json.dumps(payload).encode("utf-8")

    try:
        req = urllib.request.Request(
            URL, data=data, headers={"Content-Type": "application/json"}
        )
        with urllib.request.urlopen(req) as response:
            result = json.loads(response.read().decode("utf-8"))

            text = result["candidates"][0]["content"]["parts"][0]["text"]

            # Clean markdown
            if "```json" in text:
                text = text.split("```json")[1].split("```")[0]
            elif "```" in text:
                text = text.split("```")[1].split("```")[0]

            return json.loads(text.strip())
    except Exception as e:
        print(f"Error: {e}")
        return []


def main():
    all_questions = []

    for region in REGIONS:
        qs = generate_for_region(region)
        all_questions.extend(qs)
        print(f" -> Collected {len(qs)} questions.")
        time.sleep(1)

    if not os.path.exists("data"):
        os.makedirs("data")

    with open("data/questions.json", "w", encoding="utf-8") as f:
        json.dump(all_questions, f, ensure_ascii=False, indent=2)

    print(f"\nSUCCESS! Total questions: {len(all_questions)}")
    print("File saved to: data/questions.json")


if __name__ == "__main__":
    main()
